torch/distributed/checkpoint/state_dict_saver.py

```diff
@@ -105,21 +107,47 @@ def save(
     """
     torch._C._log_api_usage_once("torch.distributed.checkpoint.save")
 
-    dumpable_state_dict = {}
-    for key, elem in state_dict.items():
-        dumpable_state_dict[key] = (
-            elem.state_dict() if isinstance(elem, Stateful) else elem
-        )
-
     return _save_state_dict(
-        dumpable_state_dict,
+        _stateful(state_dict),
         storage_writer,
         process_group,
         coordinator_rank,
         no_dist,
         planner,
     )
 
+def async_save(state_dict: STATE_DICT_TYPE, *args: Any, **kwargs: Any) -> Future:
```

Comment:
I'm not a fan of having `*args` and `**kwargs` for a public API. This makes it harder for users to read the document and to understand how to use the args.

Reply:
I think there's an argument to be made that this is easier to maintain, but I see your point as well :) Switched to named params.

Reply:
If it is an internal API, I'm totally fine with the usage of *args and **kwargs. For the maintenance effort, we can add an UT to compare the signatures of the two APIs and raise an exception if there is a mismatch. You can achieve this with inspect.